src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1336 @@
+        radius (float): distance in raster coordinate system units which the 
+            output distance raster will be thresholded to in the future.
+            This is used to optimize the algorithm.
+        out_path (str): raster path to write out the distance results
+
+    Returns:
+        None
+    """
+    index = rtree.index.Index(interleaved=True)
+    segments = []
+    for (x1, y1), (x2, y2) in iter_linestring_segments(linestring_path):
+        segments.append(((x1, y1), (x2, y2)))
+        x_min, y_min = min(x1, x2) - radius, min(y1, y2) - radius
+        x_max, y_max = max(x1, x2) + radius, max(y1, y2) + radius
+        index.insert(len(segments) - 1, (x_min, y_min, x_max, y_max))
+
+    # Open the x_coords and y_coords rasters for reading
+    x_coord_raster = gdal.OpenEx(x_coords_path, gdal.OF_RASTER)
+    x_coord_band = x_coord_raster.GetRasterBand(1)
+    y_coord_raster = gdal.OpenEx(y_coords_path, gdal.OF_RASTER)
+    y_coord_band = y_coord_raster.GetRasterBand(1)
+    # Create and open the output raster for writing
+    pygeoprocessing.new_raster_from_base(
+        x_coords_path, out_path, gdal.GDT_Float32, [NODATA])
+    out_raster = gdal.OpenEx(out_path, gdal.OF_RASTER | gdal.GA_Update)
+    out_band = out_raster.GetRasterBand(1)
+
+    # Assuming that the x_coords and y_coords rasters have identical properties
+    raster_info = pygeoprocessing.get_raster_info(x_coords_path)
+    x_origin = raster_info['geotransform'][0]
+    y_origin = raster_info['geotransform'][3]
+    x_pixel_size, y_pixel_size = raster_info['pixel_size']
+    # iterate blockwise over the x_coords and y_coords
+    # need to use iterblocks here because we need the block offsets
+    for block in pygeoprocessing.iterblocks((x_coords_path, 1), offset_only=True):
+        x_coords = x_coord_band.ReadAsArray(**block)
+        y_coords = y_coord_band.ReadAsArray(**block)
+        # xoff, yoff are not necessarily smaller than xoff+xsize, yoff+ysize
+        # get min and max to look up in rtree index
+        x_edges = (
+            x_origin + block['xoff'] * x_pixel_size, 
+            x_origin + (block['xoff'] + block['win_xsize']) * x_pixel_size)
+        y_edges = (
+            y_origin + block['yoff'] * y_pixel_size,
+            y_origin + (block['yoff'] + block['win_ysize']) * y_pixel_size)
+        block_bbox = (
+            min(*x_edges),
+            min(*y_edges),
+            max(*x_edges),
+            max(*y_edges))
+        # Get all the line segments whose padded bounding box intersects the block
+        # These are all the line segments that could be within the radius of
+        # any point in the block.
+        line_ids = list(index.intersection(block_bbox))
+        
+        # Find each point's minimum distance to a line segment
+        min_distance_array = numpy.full(x_coords.shape, numpy.inf)
+        for line_id in line_ids:
+            (x1, y1), (x2, y2) = segments[line_id]
```

Comment:
I like how it looks as a `dict` better too. I used a list because I learned sometime long ago to never use a `dict` when you can accomplish the same thing by indexing a `list`, because it saves a little space. Happy to change it though!